torch/nativert/executor/memory/LayoutManager.cpp

```diff
@@ -157,6 +160,123 @@ void LayoutManager::populate_tensor_values() {
   }
 }
 
+#ifndef NDEBUG
+void LayoutManager::assert_no_overlapping_storages(const Node& node, size_t t)
+    const {
+  if (state_ != LayoutManagerState::Running) {
+    return;
+  }
+
+  /*
+    for each value
+    (either an input or output)
+    ensure that the associated storage
+    slice lies within the allocated slice
+    if it is managed (or if it is an alias,
+    we can use the slice allocated to its source)
+    ---
+    also ensure that the current index lies
+    within the lifetime of this value
+  */
+
+  const auto& alias_analyzer = planner_.get_alias_analyzer();
+  const auto& alive_values = alias_analyzer.alive_values_at_time(t);
+
+  // make sure active memory intervals are non-overlapping
+  // by sorting them by start, and ensuring
+  // cur.start > prev.end for each
+  std::set<std::pair<size_t, size_t>> intervals;
+
+  planner_.with_plan([&](const LayoutPlan& plan) {
+    for (auto v : alive_values) {
+      // sanity check lifetimes
+      const auto& lt = alias_analyzer.lifetime(v);
+      TORCH_CHECK_GE(t, lt.start);
+      TORCH_CHECK_LE(t, lt.end);
+
+      // prevent recomputation from occuring
+      c10::FastSet<ValueId> checked_values;
+
+      const auto& iv = parent_frame_.getIValue(v->id());
+      if (!iv.isTensor()) {
+        continue;
+      }
+
+      const auto& storage_impl = iv.toTensor().storage().unsafeGetStorageImpl();
+      const auto storage_nbytes = storage_impl->nbytes();
+
+      const auto start =
+          layout_buffer_.get_offset_from_ptr(storage_impl->data_ptr().get());
+      if (C10_UNLIKELY(!start.has_value() /* not in managed buffer */)) {
+        continue;
+      }
+
+      const auto* src_ptr = alias_analyzer.get_source_of_alias(v);
+
+      if (!src_ptr /* v isn't an alias */) {
+        if (checked_values.emplace(v->id())
+                .second /* if not already checked */) {
+          // sanity check that the storage lies within the slice allocated
+          // during planning
+          auto& alloc =
+              plan.allocations[value_to_vector_idx_map_.at(src_ptr->id())];
+          TORCH_CHECK_GE(*start, alloc.offset);
+          TORCH_CHECK_LE(*start + storage_nbytes, alloc.offset + alloc.size);
+          intervals.emplace(*start, *start + storage_nbytes);
+        }
+        continue;
+      }
+
+      const auto& src_iv = parent_frame_.getIValue(src_ptr->id());
```

Comment:
checking logic looks duplicated for iv and src_iv? 

refactor them into a function?